### PR TITLE
fix(#8539): Fix Sentinel infinite loop on failed transitions

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

- advances seq when a tombstone fails to be created
- adds failed doc transitions history. Docs that have failed transitions more than 5 times are not enqueued any more. 

medic/cht-core#8539

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8539-advance-sentinel-seq-on-failed-tombstone/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8539-advance-sentinel-seq-on-failed-tombstone/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8539-advance-sentinel-seq-on-failed-tombstone/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
Hi @garethbowen 

This is one option for keeping a history of failed transition docs. We can opt to skip this entirely, but I wanted to see how it would look like. Would you mind checking it out and sharing your thoughts? Thanks!

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -101,7 +101,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        grunt-cmd: ['ci-webdriver-default', 'ci-webdriver-standard', 'ci-e2e-integration', 'ci-e2e', 'ci-e2e-mobile', 'ci-webdriver-default-mobile']
+        grunt-cmd: ['ci-webdriver-default', 'ci-webdriver-standard', 'ci-e2e-integration', 'ci-webdriver-default-mobile']
```
Comment (garethbowen) [inline .github/workflows/build.yml:104]: todo: make this a separate commit

File: sentinel/src/lib/feed.js [hunk 2]
```diff
@@ -74,13 +78,16 @@ const changeQueue = async.queue((change, callback) => {
...
   }
   if (change.deleted) {
-    return tombstoneUtils.processChange(Promise, db.medic, change, logger)
-      .then(() => updateMetadata(change, callback))
-      .catch(callback);
+    return tombstoneUtils
+      .processChange(Promise, db.medic, change, logger)
+      .catch(() => {}) // error already logged by tombstoneUtils
+      // advance sentinel seq even when tombstone creation fails
+      .then(() => updateMetadata(change, callback));
```
Comment (garethbowen) [inline sentinel/src/lib/feed.js:85]: todo: make this a separate commit

File: sentinel/src/lib/change-retry-history.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+const shouldProcess = (change) => {
+  const key = getKey(change);
+
+  return !history[key] || history[key] <= MAX_RETRIES;
+};
+
+module.exports = {
+  add,
+  shouldProcess,
+};
```
Comment (garethbowen) [inline sentinel/src/lib/change-retry-history.js:36]: You know how much I love caches...

... but I think this is ok.

An alternative would be to only detect when the transition seq isn't advancing at all. That is to say if an entire batch is processed and 100% fails 5 times in a row then skip over them and move on. This way you're only caching 1 number instead of 1000 keys. Simpler and therefore less buggy?

One advantage of your approach is if you had 99 failing docs my approach would never get past them but yours would. 

Are there other advantages of your suggestion?

Comment (dianabarsan) [reply sentinel/src/lib/change-retry-history.js:36]: I'll check to see how your suggestion looks like. I think it requires adding some conditionals in the `feed.js` code that would increase complexity.

Comment (dianabarsan) [reply sentinel/src/lib/change-retry-history.js:36]: There's no guarantee that requesting changes from a certains seq will return the same results every time (because shards and because docs can get updated). 
So saying that a "batch" failed 5 times, it would still mean we need to check every doc in that batch and store which docs were in the failed batch. So we'd still be storing 100 changes in memory, along with the seq. 

Do you think this is worth pursuing and would be easier to maintain or understand than straight up caching doc tries without the seq?

Comment (garethbowen) [reply sentinel/src/lib/change-retry-history.js:36]: Yeah that's a good point. Go with it as is.
